rust-runtime/aws-smithy-runtime/src/client/retries/token_bucket.rs

```diff
@@ -77,10 +141,40 @@ impl TokenBucket {
     }
 
     pub(crate) fn regenerate_a_token(&self) {
-        if self.semaphore.available_permits() < self.max_permits {
-            trace!("adding {PERMIT_REGENERATION_AMOUNT} back into the bucket");
-            self.semaphore.add_permits(PERMIT_REGENERATION_AMOUNT)
+        self.add_tokens(PERMIT_REGENERATION_AMOUNT);
+    }
+
+    pub(crate) fn reward_success(&self) {
+        let mut calc_fractional_tokens = self.fractional_tokens.load();
+        // Verify that fractional tokens have not become corrupted
+        if !calc_fractional_tokens.is_finite() {
+            tracing::error!("Fractional tokens corrupted to: {}", calc_fractional_tokens);
+            // If corrupted, reset to the number of permits the bucket was created with
+            self.fractional_tokens.store(self.max_permits as f32);
```

Comment:
I still think this is "wrong", we're rewarding success and every whole token is converted and added to our actual tokens/permits. Setting this to max permits is just going to immediately on the next go around fill up the bucket to max permits. I think this should be zero.